viewer-prototype/src/browser/trace-viewer/components/utils/filter-tree-component.tsx

```diff
@@ -0,0 +1,230 @@
+import * as React from 'react';
+import { Entry } from 'tsp-typescript-client/lib/models/entry';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faChevronDown, faChevronRight } from '@fortawesome/free-solid-svg-icons'
+import { CheckboxComponent } from './checkbox-component';
+
+type FilterTreeProps = {
+    nodes: Entry[];
+    checkedSeries: number[];
+    onChecked: (ids: number[]) => void;
+}
+
+type FilterTreeState = {
+    treeNodes: TreeNode[];
+    checkedIDs: number[];
+}
+
+export class FilterTree extends React.Component<FilterTreeProps, FilterTreeState> {
+    constructor(props: FilterTreeProps) {
+        super(props);
+        this.state = {
+            treeNodes: listToTree(this.props.nodes),
+            checkedIDs: this.props.checkedSeries
+        }
+    }
+
+    shouldComponentUpdate(nextProps: FilterTreeProps, nextState: FilterTreeState) {
+        return (this.props.checkedSeries !== nextProps.checkedSeries || this.props.nodes !== nextProps.nodes)
+        
+    }
+
+    componentWillReceiveProps(nextProps: FilterTreeProps) {
+        this.setState({
+            treeNodes: listToTree(nextProps.nodes),
+            checkedIDs: nextProps.checkedSeries
+        })
+    }
+
+    getRootNodes = () => {
+        const nodes = this.state.treeNodes;
+        return nodes.filter((node: TreeNode) => node.isRoot === true);
+    }
+
+    getChildNodes = (node: TreeNode) => {
```

Comment:
While this is a perfectly legit typescript statement, I wonder if there should be a return type somewhere, like getChildNodes: something?